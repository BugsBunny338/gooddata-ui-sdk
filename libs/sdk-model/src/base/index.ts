// (C) 2019 GoodData Corporation

import isEmpty = require("lodash/isEmpty");

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type Identifier = string;

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export interface IObjUriQualifier {
    uri: string;
}

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export interface IObjIdentifierQualifier {
    identifier: string;
}

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export interface IObjLocalIdentifierQualifier {
    localIdentifier: Identifier;
}

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type ObjQualifier = IObjUriQualifier | IObjIdentifierQualifier;

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type ObjQualifierWithLocal = ObjQualifier | IObjLocalIdentifierQualifier;

//
// Type guards
//

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export function isUriQualifier(obj: any): obj is IObjUriQualifier {
    return !isEmpty(obj) && (obj as IObjUriQualifier).uri !== undefined;
}

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export function isIdentifierQualifier(obj: any): obj is IObjIdentifierQualifier {
    return !isEmpty(obj) && (obj as IObjIdentifierQualifier).identifier !== undefined;
}
