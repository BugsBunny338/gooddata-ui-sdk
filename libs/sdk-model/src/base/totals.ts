// (C) 2019 GoodData Corporation

import isEmpty from "lodash/isEmpty";
/**
 * TODO: SDK8: Add docs
 *
 * @public
 */

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export type TotalType = "sum" | "avg" | "max" | "min" | "med" | "nat";

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export interface ITotal {
    type: TotalType;
    measureIdentifier: string;
    attributeIdentifier: string;
    alias?: string;
}

//
// Type guards
//

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export function isTotal(obj: any): obj is ITotal {
    return !isEmpty(obj) && (obj as ITotal).type !== undefined;
}

//
//
//

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export function totalIsNative(total: ITotal): boolean {
    return total.type === "nat";
}
