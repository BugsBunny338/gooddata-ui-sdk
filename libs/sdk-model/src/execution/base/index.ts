// (C) 2019-2020 GoodData Corporation

import isEmpty = require("lodash/isEmpty");
import invariant from "ts-invariant";

/**
 * Type for all identifiers.
 *
 * @public
 */
export type Identifier = string;

/**
 * Type for all URI's.
 *
 * @public
 */
export type Uri = string;

/**
 * Model object reference using object's URI.
 *
 * NOTE: using URI references is discouraged. URIs are workspace-specific and thus any application
 * which references model objects using their URI will not work on multiple workspaces.
 *
 * @public
 */
export type UriRef = {
    uri: Uri;
};

/**
 * Model object reference using object's unique identifier.
 *
 * NOTE: this is preferred way to reference model objects.
 *
 * @public
 */
export type IdentifierRef = {
    identifier: Identifier;
};

/**
 * Model object reference using object's local identifier. This type of referencing can be used for objects
 * that are defined together within the same scope - such as within same execution.
 *
 * @public
 */
export type LocalIdRef = {
    localIdentifier: Identifier;
};

/**
 * Model object reference. Objects can be referenced by their URI or by identifier.
 *
 * NOTE: using URI references is discouraged. URIs are workspace-specific and thus any application
 * which references model objects using their URI will not work on multiple workspaces.
 *
 * @public
 */
export type ObjRef = UriRef | IdentifierRef;

/**
 * Model object reference with support of referencing objects living in the same scope using their
 * local identifier.
 *
 * @public
 */
export type ObjRefInScope = ObjRef | LocalIdRef;

//
// Type guards
//

/**
 * Type guard checking whether object is an URI Reference.
 *
 * @public
 */
export function isUriRef(obj: any): obj is UriRef {
    return !isEmpty(obj) && (obj as UriRef).uri !== undefined;
}

/**
 * Type guard checking whether object is an Identifier Reference.
 *
 * @public
 */
export function isIdentifierRef(obj: any): obj is IdentifierRef {
    return !isEmpty(obj) && (obj as IdentifierRef).identifier !== undefined;
}

/**
 * Type guard checking whether object is an Identifier Reference or an URI reference.
 *
 * @public
 */
export function isObjRef(obj: any): obj is ObjRef {
    return isUriRef(obj) || isIdentifierRef(obj);
}

/**
 * Type guard checking whether object is a localId Reference.
 *
 * @public
 */
export function isLocalIdRef(obj: any): obj is LocalIdRef {
    return !isEmpty(obj) && (obj as LocalIdRef).localIdentifier !== undefined;
}

/**
 * Retrieves value of object reference.
 *
 * @public
 */
export function objectRefValue(objRef: ObjRef | ObjRefInScope): string {
    invariant(objRef, "object reference must not be undefined");

    if (isIdentifierRef(objRef)) {
        return objRef.identifier;
    } else if (isUriRef(objRef)) {
        return objRef.uri;
    }

    return objRef.localIdentifier;
}

/**
 * Returns a value indicating whether the two ObjRef instances are semantically equal (i.e. are of the same type and have the same value).
 * Null and undefined are considered equal to each other.
 *
 * @public
 */
export function areObjRefsEqual(
    a: ObjRefInScope | null | undefined,
    b: ObjRefInScope | null | undefined,
): boolean {
    if (a == null) {
        return b == null;
    }
    if (isIdentifierRef(a)) {
        return isIdentifierRef(b) && a.identifier === b.identifier;
    }
    if (isUriRef(a)) {
        return isUriRef(b) && a.uri === b.uri;
    }
    return isLocalIdRef(b) && a.localIdentifier === b.localIdentifier;
}
