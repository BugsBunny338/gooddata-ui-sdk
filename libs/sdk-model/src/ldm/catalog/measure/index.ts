// (C) 2019-2020 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import { IGroupableCatalogItemBase } from "../group";
import { IMeasureMetadataObject } from "../../metadata/measure";

/**
 * Type representing catalog attribute
 *
 * @public
 */
export interface ICatalogMeasure extends IGroupableCatalogItemBase {
    type: "measure";
    measure: IMeasureMetadataObject;
}

/**
 * Type guard checking whether the provided object is a {@link ICatalogMeasure}
 *
 * @public
 */
export function isCatalogMeasure(obj: any): obj is ICatalogMeasure {
    return !isEmpty(obj) && (obj as ICatalogMeasure).type === "measure";
}
