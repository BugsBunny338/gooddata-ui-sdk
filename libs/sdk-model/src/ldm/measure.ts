// (C) 2019-2020 GoodData Corporation
import { ObjectType } from "../objRef";

/**
 * Token representing part of parsed MAQL measure expression
 * e.g. "SELECT [/gdc/md/projectId/obj/6273] WHERE [/gdc/md/projectId/obj/6307] = [/gdc/md/projectId/obj/6307/elements?id=5703453]"
 * could be represented as:
 *  \[
 *      \{
 *          "type": "text",
 *          "value": "SELECT "
 *      \},
 *      \{
 *          "type": "measure",
 *          "value": "Amount",
 *      \},
 *      \{
 *          "type": "text",
 *          "value": " WHERE "
 *      \},
 *      \{
 *          "type": "attribute",
 *          "value": "Status",
 *      \},
 *      \{
 *          "type": "text",
 *          "value": " = "
 *      \},
 *      \{
 *          "type": "attributeElement",
 *          "value": "Won",
 *      \}
 *  \]
 *
 * @public
 */
export type IMeasureExpressionToken =
    | IObjectExpressionToken
    | IAttributeElementExpressionToken
    | ITextExpressionToken;

/**
 * Parsed maql token referencing metadata object
 *
 * @public
 */
export interface IObjectExpressionToken {
    /**
     * Expression token type
     */
    type: ObjectType;

    /**
     * Title of the object
     */
    value: string;
}

/**
 * Parsed maql token referencing attribute element
 *
 * @public
 */
export interface IAttributeElementExpressionToken {
    /**
     * Expression token type
     */
    type: "attributeElement";

    /**
     * Element title
     */
    value: string | undefined;

    /**
     * Deleted element
     */
    deleted?: boolean;
}

/**
 * Parsed maql text
 *
 * @public
 */
export interface ITextExpressionToken {
    /**
     * Expression token type
     */
    type: "text";

    /**
     * Plain text
     */
    value: string;
}
