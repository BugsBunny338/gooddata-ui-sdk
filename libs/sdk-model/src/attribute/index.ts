// (C) 2019 GoodData Corporation
import { Identifier, ObjQualifier } from "../base";
import isEmpty from "lodash/isEmpty";

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export interface IAttribute {
    attribute: {
        localIdentifier: Identifier;
        displayForm: ObjQualifier;
        alias?: string;
    };
}

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export function isAttribute(obj: any): obj is IAttribute {
    return !isEmpty(obj) && (obj as IAttribute).attribute !== undefined;
}

//
//
//

/**
 * TODO: SDK8: Add docs
 *
 * @public
 */
export function attributeId(a: IAttribute): string {
    return a.attribute.localIdentifier;
}
