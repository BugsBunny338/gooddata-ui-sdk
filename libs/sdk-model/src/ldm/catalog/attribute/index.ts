// (C) 2019-2020 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import { IAttributeMetadataObject } from "../../metadata/attribute";
import { IAttributeDisplayFormMetadataObject } from "../../metadata/attributeDisplayForm";
import { IGroupableCatalogItemBase } from "../group";

/**
 * Type representing catalog attribute
 *
 * @public
 */
export interface ICatalogAttribute extends IGroupableCatalogItemBase {
    /**
     * Catalog item type
     */
    type: "attribute";

    /**
     * Attribute metadata object that catalog attribute represents
     */
    attribute: IAttributeMetadataObject;

    /**
     * Default display form of the attribute
     */
    defaultDisplayForm: IAttributeDisplayFormMetadataObject;
}

/**
 * Type guard checking whether the provided object is a {@link ICatalogAttribute}
 *
 * @public
 */
export function isCatalogAttribute(obj: any): obj is ICatalogAttribute {
    return !isEmpty(obj) && (obj as ICatalogAttribute).type === "attribute";
}
