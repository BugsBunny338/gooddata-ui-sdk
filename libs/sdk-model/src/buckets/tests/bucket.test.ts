// (C) 2019 GoodData Corporation

import {
    applyRatioRule,
    AttributeOrMeasure,
    bucketAttribute,
    bucketAttributes,
    bucketIsEmpty,
    bucketItems,
    bucketMeasure,
    bucketMeasures,
    bucketTotals,
    ComputeRatioRule,
    newBucket,
} from "../index";
import { Account, Activity, Velocity, Won } from "../../../__mocks__/model";
import { InvariantError } from "ts-invariant";
import { ITotal, newTotal } from "../../base/totals";
import { attributeLocalId, IAttribute, idMatchAttribute } from "../../attribute";
import { idMatchMeasure, IMeasure, measureId } from "../../measure";
import { modifyMeasure } from "../..";

describe("newBucket", () => {
    const Scenarios: Array<[string, any, any[]]> = [
        ["empty bucket", "emptyBucket", [undefined]],
        ["bucket with single measure", "measureBucket", [Won]],
        ["bucket with multiple measures", "measureBucket", [Won, Velocity.Sum]],
        ["bucket with single attribute", "attributeBucket", [Account.Default]],
        ["bucket with multiple attributes", "attributeBucket", [Account.Default, Account.Name]],
        [
            "bucket with mix of attributes and measures",
            "mixedBucket",
            [Won, Account.Name, Velocity.Sum, Account.Default],
        ],
        ["bucket with totals", "totalsBucket", [Won, Account.Name, newTotal("sum", Won, Account.Name)]],
        ["bucket with totals when items not in bucket", "totalsBucket", [newTotal("sum", Won, Account.Name)]],
    ];

    it.each(Scenarios)("should create %s", (_desc, localIdArg, contentArg) => {
        expect(newBucket(localIdArg, ...contentArg)).toMatchSnapshot();
    });

    it("should throw if local identifier not specified", () => {
        expect(() => newBucket(undefined as any, Won, Account.Name)).toThrowError(InvariantError);
    });

    it("should throw if content is invalid type", () => {
        expect(() => newBucket("invalidBucket", Won, {} as any, Account.Name)).toThrowError(InvariantError);
    });
});

const Total = newTotal("sum", Won, Account.Name);
const EmptyBucket = newBucket("EmptyBucket");
const BucketWithOnlyTotals = newBucket("BucketWithOnlyTotals", newTotal("sum", Won, Account.Name));
const BucketWithSingleMeasure = newBucket("BucketWithSingleMeasure", Won);
const BucketWithManyMeasures = newBucket("BucketWithManyMeasures", Won, Velocity.Sum, Velocity.Max);
const BucketWithSingleAttr = newBucket("BucketWithSingleAttr", Account.Name);
const BucketWithManyAttrs = newBucket("BucketWithManyAttrs", Account.Name, Account.Default, Activity.Subject);
const BucketWithMeasureAndAttr = newBucket("BucketWithMeasureAndAttr", Won, Account.Name);
const BucketWithEverything = newBucket("BucketWithMeasureAndAttr", Won, Account.Name, Total);

describe("bucketIsEmpty", () => {
    const Scenarios: Array<[boolean, string, any]> = [
        [true, "empty bucket", EmptyBucket],
        [true, "undefined input", undefined],
        [true, "null input", null],
        [false, "bucket with only totals", BucketWithOnlyTotals],
        [false, "bucket with single measure", BucketWithSingleMeasure],
        [false, "bucket with single attr", BucketWithSingleAttr],
        [false, "bucket with all possible inputs", BucketWithEverything],
    ];

    it.each(Scenarios)("should return %s for %s", (expectedResult, _desc, input) => {
        expect(bucketIsEmpty(input)).toEqual(expectedResult);
    });
});

describe("bucketAttribute", () => {
    const Scenarios: Array<[string, any, any, IAttribute | undefined]> = [
        ["no attribute in empty bucket", EmptyBucket, undefined, undefined],
        ["no attribute for undefined bucket", undefined, undefined, undefined],
        ["no attribute for null bucket", null, undefined, undefined],
        ["no attribute in measure-only bucket", BucketWithManyMeasures, undefined, undefined],
        ["no attribute if localId empty", BucketWithManyAttrs, "", undefined],
        ["no attribute if localId matches nothing", BucketWithManyAttrs, "noSuchLocalId", undefined],
        ["first attribute when no predicate provided", BucketWithManyAttrs, undefined, Account.Name],
        ["first attribute in mixed attr & measure bucket", BucketWithMeasureAndAttr, undefined, Account.Name],
        [
            "attribute by local id if predicate is a string",
            BucketWithManyAttrs,
            attributeLocalId(Account.Default),
            Account.Default,
        ],
        ["no attribute if no predicate match", BucketWithManyAttrs, () => false, undefined],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketAttribute(bucketArg, predicateArg)).toEqual(expectedResult);
    });
});

describe("bucketAttributes", () => {
    const Scenarios: Array<[string, any, any, IAttribute[]]> = [
        ["no attributes in empty bucket", EmptyBucket, undefined, []],
        ["no attributes if bucket undefined", undefined, undefined, []],
        ["no attributes if bucket null", null, undefined, []],
        ["no attributes in measure-only bucket", BucketWithManyMeasures, undefined, []],
        ["no attributes in attr bucket but no predicate match", BucketWithManyAttrs, () => false, []],
        ["attributes in single attr bucket", BucketWithSingleAttr, undefined, [Account.Name]],
        [
            "attributes in many attr bucket",
            BucketWithManyAttrs,
            undefined,
            [Account.Name, Account.Default, Activity.Subject],
        ],
        [
            "single attribute matching predicate",
            BucketWithManyAttrs,
            idMatchAttribute(attributeLocalId(Account.Name)),
            [Account.Name],
        ],
        ["attributes in mixed bucket", BucketWithMeasureAndAttr, undefined, [Account.Name]],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketAttributes(bucketArg, predicateArg)).toEqual(expectedResult);
    });
});

describe("bucketMeasure", () => {
    const Scenarios: Array<[string, any, any, IMeasure | undefined]> = [
        ["no measure in empty bucket", EmptyBucket, undefined, undefined],
        ["no measure for undefined bucket", undefined, undefined, undefined],
        ["no measure for null bucket", null, undefined, undefined],
        ["no measure in attr-only bucket", BucketWithManyAttrs, undefined, undefined],
        ["no measure if localId empty", BucketWithManyMeasures, "", undefined],
        ["no measure if localId matches nothing", BucketWithManyMeasures, "noSuchLocalId", undefined],
        ["first measure when no predicate provided", BucketWithManyMeasures, undefined, Won],
        ["first measure in mixed attr & measure bucket", BucketWithMeasureAndAttr, undefined, Won],
        [
            "measure by local id if predicate is a string",
            BucketWithManyMeasures,
            measureId(Velocity.Sum),
            Velocity.Sum,
        ],
        ["no measure if no predicate match", BucketWithManyMeasures, () => false, undefined],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketMeasure(bucketArg, predicateArg)).toEqual(expectedResult);
    });
});

describe("bucketMeasures", () => {
    const Scenarios: Array<[string, any, any, IMeasure[]]> = [
        ["no measures in empty bucket", EmptyBucket, undefined, []],
        ["no measures if bucket undefined", undefined, undefined, []],
        ["no measures if bucket null", null, undefined, []],
        ["no measures in attr-only bucket", BucketWithManyAttrs, undefined, []],
        ["no measures in measure bucket but no predicate match", BucketWithManyMeasures, () => false, []],
        ["measures in single measure bucket", BucketWithSingleMeasure, undefined, [Won]],
        [
            "measures in many measure bucket",
            BucketWithManyMeasures,
            undefined,
            [Won, Velocity.Sum, Velocity.Max],
        ],
        [
            "single measure matching predicate",
            BucketWithManyMeasures,
            idMatchMeasure(measureId(Velocity.Sum)),
            [Velocity.Sum],
        ],
        ["measures in mixed bucket", BucketWithMeasureAndAttr, undefined, [Won]],
    ];

    it.each(Scenarios)("should find %s", (_desc, bucketArg, predicateArg, expectedResult) => {
        expect(bucketMeasures(bucketArg, predicateArg)).toEqual(expectedResult);
    });
});

describe("bucketItems", () => {
    const Scenarios: Array<[string, any, AttributeOrMeasure[]]> = [
        ["no items in empty bucket", EmptyBucket, []],
        ["no items if bucket undefined", undefined, []],
        ["no items if bucket null", null, []],
        ["items in populated bucket", BucketWithMeasureAndAttr, [Won, Account.Name]],
    ];

    it.each(Scenarios)("should return %s", (_desc, bucketArg, expectedResult) => {
        expect(bucketItems(bucketArg)).toEqual(expectedResult);
    });
});

describe("bucketTotals", () => {
    const Scenarios: Array<[string, any, ITotal[]]> = [
        ["no totals in empty bucket", EmptyBucket, []],
        ["no totals in bucket without totals", BucketWithMeasureAndAttr, []],
        ["no items if bucket undefined", undefined, []],
        ["no items if bucket null", null, []],
        ["totals in populated bucket", BucketWithEverything, [Total]],
    ];

    it.each(Scenarios)("should return %s", (_desc, bucketArg, expectedResult) => {
        expect(bucketTotals(bucketArg)).toEqual(expectedResult);
    });
});

describe("applyRatioRule", () => {
    const MeasureWithRatio1 = modifyMeasure(Won, m => m.ratio());
    const MeasureWithRatio2 = modifyMeasure(Velocity.Avg, m => m.ratio());

    const Scenarios: Array<[string, any, any, any]> = [
        ["empty array if undefined input", undefined, undefined, []],
        ["empty array if null input", null, undefined, []],
        [
            "apply SINGLE rule by default",
            [MeasureWithRatio1],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [MeasureWithRatio1],
        ],
        [
            "keep all ratios if rule is ALWAYS",
            [MeasureWithRatio1, MeasureWithRatio2],
            ComputeRatioRule.ANY_MEASURE,
            [MeasureWithRatio1, MeasureWithRatio2],
        ],
        [
            "reset all ratios if rule is NEVER",
            [MeasureWithRatio1, MeasureWithRatio2],
            ComputeRatioRule.NEVER,
            [Won, Velocity.Avg],
        ],
        [
            "reset all ratios if rule is SINGLE and there are two ratio measures",
            [MeasureWithRatio1, MeasureWithRatio2],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [Won, Velocity.Avg],
        ],
        [
            "keep ratio of single ratio measure",
            [MeasureWithRatio1],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [MeasureWithRatio1],
        ],
        [
            "keep ratio of single ratio measure when mixed with attributes",
            [MeasureWithRatio1, Account.Name],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [MeasureWithRatio1, Account.Name],
        ],
        [
            "retain attributes as-is when SINGLE rule",
            [Activity.Subject, MeasureWithRatio1, Account.Name],
            ComputeRatioRule.SINGLE_MEASURE_ONLY,
            [Activity.Subject, MeasureWithRatio1, Account.Name],
        ],
        [
            "retain attributes as-is when NEVER rule",
            [Activity.Subject, MeasureWithRatio1, Account.Name],
            ComputeRatioRule.NEVER,
            [Activity.Subject, Won, Account.Name],
        ],
    ];

    it.each(Scenarios)("should return %s", (_desc, itemsArg, ruleArg, expectedResult) => {
        expect(applyRatioRule(itemsArg, ruleArg)).toEqual(expectedResult);
    });
});
