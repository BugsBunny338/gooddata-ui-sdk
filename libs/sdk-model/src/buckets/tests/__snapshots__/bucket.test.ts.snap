// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`newBucket should create bucket with mix of attributes and measures 1`] = `
Object {
  "items": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.account.id.name",
        },
        "localIdentifier": "a_label.account.id.name",
      },
    },
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "aggregation": "sum",
            "item": Object {
              "identifier": "fact.stagehistory.velocity",
            },
          },
        },
        "localIdentifier": "m_fact.stagehistory.velocity_sum",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.account.id",
        },
        "localIdentifier": "a_label.account.id",
      },
    },
  ],
  "localIdentifier": "mixedBucket",
}
`;

exports[`newBucket should create bucket with multiple attributes 1`] = `
Object {
  "items": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.account.id",
        },
        "localIdentifier": "a_label.account.id",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.account.id.name",
        },
        "localIdentifier": "a_label.account.id.name",
      },
    },
  ],
  "localIdentifier": "attributeBucket",
}
`;

exports[`newBucket should create bucket with multiple measures 1`] = `
Object {
  "items": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "aggregation": "sum",
            "item": Object {
              "identifier": "fact.stagehistory.velocity",
            },
          },
        },
        "localIdentifier": "m_fact.stagehistory.velocity_sum",
      },
    },
  ],
  "localIdentifier": "measureBucket",
}
`;

exports[`newBucket should create bucket with single attribute 1`] = `
Object {
  "items": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.account.id",
        },
        "localIdentifier": "a_label.account.id",
      },
    },
  ],
  "localIdentifier": "attributeBucket",
}
`;

exports[`newBucket should create bucket with single measure 1`] = `
Object {
  "items": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
  ],
  "localIdentifier": "measureBucket",
}
`;

exports[`newBucket should create bucket with totals 1`] = `
Object {
  "items": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "afSEwRwdbMeQ",
            },
          },
        },
        "localIdentifier": "m_afSEwRwdbMeQ",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.account.id.name",
        },
        "localIdentifier": "a_label.account.id.name",
      },
    },
  ],
  "localIdentifier": "totalsBucket",
  "totals": Array [
    Object {
      "attributeIdentifier": "a_label.account.id.name",
      "measureIdentifier": "m_afSEwRwdbMeQ",
      "type": "sum",
    },
  ],
}
`;

exports[`newBucket should create bucket with totals when items not in bucket 1`] = `
Object {
  "items": Array [],
  "localIdentifier": "totalsBucket",
  "totals": Array [
    Object {
      "attributeIdentifier": "a_label.account.id.name",
      "measureIdentifier": "m_afSEwRwdbMeQ",
      "type": "sum",
    },
  ],
}
`;

exports[`newBucket should create empty bucket 1`] = `
Object {
  "items": Array [],
  "localIdentifier": "emptyBucket",
}
`;
